interop/lib/src/client.dart

```diff
@@ -0,0 +1,1027 @@
+  ///   if using a usable auth implementation, it may specify the file location
+  ///   in the environment variable GOOGLE_APPLICATION_CREDENTIALS
+  /// * optionally uses the flag `--oauth_scope` for the oauth scope if
+  ///   implementator wishes to use service account credential instead of JWT
+  ///   credential. For testing against grpc-test.sandbox.googleapis.com, oauth
+  ///   scope "https://www.googleapis.com/auth/xapi.zoo" should be used.
+  ///
+  /// Procedure:
+  /// 1. Client configures the channel with just SSL credentials
+  /// 2. Client calls UnaryCall, setting per-call credentials to
+  ///    JWTTokenCredentials. The request is the following message
+  ///     {
+  ///       fill_username: true
+  ///     }
+  ///
+  /// Client asserts:
+  /// * call was successful
+  /// * received SimpleResponse.username is not empty and is in the json key
+  ///   file used by the auth library. The client can optionally check the
+  ///   username matches the email address in the key file.
+  Future<Null> perRpcCreds() async {
+    throw 'Not implemented';
+  }
+
+  /// This test verifies that custom metadata in either binary or ascii format
+  /// can be sent as initial-metadata by the client and as both initial- and
+  /// trailing-metadata by the server.
+  ///
+  /// Procedure:
+  /// 1. The client attaches custom metadata with the following keys and values:
+  ///     key: "x-grpc-test-echo-initial", value: "test_initial_metadata_value"
+  ///     key: "x-grpc-test-echo-trailing-bin", value: 0xababab
+  ///    to a UnaryCall with request:
+  ///     {
+  ///       response_size: 314159
+  ///       payload: {
+  ///         body: 271828 bytes of zeros
+  ///       }
+  ///     }
+  ///
+  /// 2. The client attaches custom metadata with the following keys and values:
+  ///     key: "x-grpc-test-echo-initial", value: "test_initial_metadata_value"
+  ///     key: "x-grpc-test-echo-trailing-bin", value: 0xababab
+  ///    to a FullDuplexCall with request:
+  ///     {
+  ///       response_parameters: {
+  ///         size: 314159
+  ///       }
+  ///       payload: {
+  ///         body: 271828 bytes of zeros
+  ///       }
+  ///     }
+  ///    and then half-closes
+  ///
+  /// Client asserts:
+  /// * call was successful
+  /// * metadata with key `"x-grpc-test-echo-initial"` and value
+  ///   `"test_initial_metadata_value"`is received in the initial metadata for
+  ///   calls in Procedure steps 1 and 2.
+  /// * metadata with key `"x-grpc-test-echo-trailing-bin"` and value `0xababab`
+  ///   is received in the trailing metadata for calls in Procedure steps 1 and
+  ///   2.
+  Future<Null> customMetadata() async {
+    void validate(Map<String, String> headers, Map<String, String> trailers) {
+      if (headers[_headerEchoKey] != _headerEchoData) {
+        throw 'Invalid header data recived.';
+      }
+      if (trailers[_trailerEchoKey] != _trailerEchoData) {
```

Comment:
[puzzled] The trailer data is not really binary, but a hex string. How is that any different from the ascii header data?

Reply:
It isn't really, but some languages have APIs to set/get binary values, which automatically append/strip the '-bin' suffix and en/decode the binary data as base64.

And I just forgot to base64 encode the hex value in the test description. Fixed. It's still just a string, though, which is why the test passed...